lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,53 @@
+class LRUHash
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?, :keys, :clear, :delete
+
+  attr_reader :table, :minimum_time_in_lru
+  def initialize(minimum_time_in_lru = 300)
+    @minimum_time_in_lru = minimum_time_in_lru
+    @table = {}
+    @lock = Mutex.new
+  end
+
+  def set(key, value)
+    @lock.synchronize {
+      @table.delete(key)
+      @table[key] = value
+      clear_unused_resources
+    }
+    value
+  end
+
+  def get(key)
+    @lock.synchronize {
+      if found = @table.delete(key)
+        @table[key] = found
+      end
+    }
+    @table[key]
+  end
+
+  def []=(key, value)
+    set(key, value)
+  end
+
+  def [](key)
+    get(key)
+  end
+
+  private
+
+  def clear_unused_resources
```

Comment:
For performance, I'd suggest you don't claim this within the synchronize, but instead claim the mutex around `#delete` yourself here.

Reply:
add comment here about what it's doing